### PR TITLE
Add app diffs to compare view

### PR DESCRIPTION
**Pull request by proteusvacuum**
Adds diffs to the app summary.

There are no links to this page yet so that we can run UAT / QA on prod.
Screenshot from 2019-04-24 13-21-27

FF: none, but hidden link.

### CODE REVIEW

**Comment by snopoke** [comment]
wow, this looks amazing! 2 points:
* having the removals crossed out makes it hard to read them
* I couldn't figure out what the "+ Realy big truck!" addition was

**Comment by proteusvacuum** [comment]
@snopoke the "Really big truck" is an added comment to that question. I've moved the comments to have an icon in front of them because just having muted text is hard to make out.

File: corehq/apps/app_manager/app_schemas/form_metadata.py [hunk 1]
```diff
@@ -123,3 +130,87 @@ def get_app_summary_formdata(domain, app, include_shadow_forms=True):
...
+        self.first = get_app_summary_formdata(app1.domain, app1)[0]
+        self.second = get_app_summary_formdata(app2.domain, app2)[0]
+
+        self._populate_id_caches()
+        self._mark_removed_items()
+        self._mark_added_items()
+
+    def _populate_id_caches(self):
+        self._first_ids = set()
+        self._first_questions_by_id = defaultdict(dict)
```
Comment (orangejenny) [inline corehq/apps/app_manager/app_schemas/form_metadata.py]: Nitpick: how about `_first_questions_by_id` => `_first_questions_by_form_id`?

File: corehq/apps/app_manager/app_schemas/form_metadata.py [hunk 2]
```diff
@@ -123,3 +130,87 @@ def get_app_summary_formdata(domain, app, include_shadow_forms=True):
...
+        for module in self.second:
+            self._second_ids.add(module['id'])
+            for form in module['forms']:
+                self._second_ids.add(form['id'])
+                for question in form['questions']:
+                    self._second_questions_by_id[form['id']][question['value']] = question
+
+    def _mark_removed_items(self):
+        for module in self.first:
+            if module['id'] not in self._second_ids:
```
Comment (orangejenny) [inline corehq/apps/app_manager/app_schemas/form_metadata.py]: Confused: `id` is the index, right? Shouldn't this look at unique id? It seems like the only way you'd mark a module as removed is if one app has a different number of modules than the other.

File: corehq/apps/app_manager/app_schemas/form_metadata.py [hunk 3]
```diff
@@ -123,3 +130,87 @@ def get_app_summary_formdata(domain, app, include_shadow_forms=True):
...
+            if question_path not in self._first_questions_by_id[form_id]:
+                second_question['diff_state'] = ADDED
+            else:
+                first_question = self._first_questions_by_id[form_id][question_path]
+                self._mark_changed_questions(first_question, second_question)
+
+    def _mark_changed_questions(self, first_question, second_question):
+        for attribute in INTERESTING_ATTRIBUTES:
+            attribute_changed = first_question.get(attribute) != second_question.get(attribute)
+            attribute_added = not first_question.get(attribute) and second_question.get(attribute)
```
Comment (orangejenny) [inline corehq/apps/app_manager/app_schemas/form_metadata.py]: Isn't this redundant with the line above? `None != (not None)` is true.

Comment (orangejenny) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: Nevermind, I see this logic gets more complicated in later commits.

File: corehq/apps/app_manager/templates/app_manager/form_summary_base.html [hunk 4]
```diff
@@ -63,7 +63,7 @@ <h5>
                 </ul>
                 <!-- /ko -->
                 <ol data-bind="foreach: questions">
-                    <li data-bind="visible: isVisible, attr: { class: diffClass }">
+                    <li data-bind="visible: isVisible, attr: { class: attributeDiff['label'] ? attributeDiff['label'] : '' }">
```
Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/form_summary_base.html]: These expressions would be more concise as `attributeDiff['label'] || ''`. Do you know if you just did `attributeDiff['label']`, would knockout insert the string "undefined" as the class?

File: corehq/apps/app_manager/app_schemas/app_case_metadata.py [hunk 5]
```diff
@@ -290,9 +290,16 @@ def editable(self):
         return vtype['editable']
 
 
+class LoadSaveProperty(JsonObject):
```
Comment (orangejenny) [inline corehq/apps/app_manager/app_schemas/app_case_metadata.py]: This is a beautiful improvement.

File: corehq/apps/app_manager/templates/app_manager/form_summary_base.html (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/app_manager/templates/app_manager/form_summary_base.html]: (realized this is outdated, removing comment)

File: corehq/apps/app_manager/static/app_manager/js/summary/form_models.js [hunk 6]
```diff
@@ -52,6 +52,16 @@ hqDefine('app_manager/js/summary/form_models',[
                     });
                 });
             },
+            onSelectChangesOnlyMenuItem: function () {
+                _.each(viewModels, function (viewModel) {
+                    _.each(viewModel.modules, function (module) {
+                        module.isSelected(module.changes.contains_changes || _.find(module.forms, function (f) { return f.changes.contains_changes; }));
+                        _.each(module.forms, function (form) {
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/summary/form_models.js]: Is it worth putting this loop inside of an `if (module.isSelected())`?

**Review by orangejenny — APPROVED** [review]
This is great.

File: corehq/apps/app_manager/app_schemas/form_metadata.py (hunk 2 above)
Comment (proteusvacuum) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: `id` is actually the `unique_id`... 
https://github.com/dimagi/commcare-hq/blob/fr%2Fform-diffs/corehq/apps/app_manager/app_schemas/form_metadata.py#L125
Maybe that is worth changing here? It's been this way in the app summary for ever, but could be worth a refactor for consistency sake.

File: corehq/apps/app_manager/app_schemas/form_metadata.py (hunk 1 above)
Comment (proteusvacuum) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: :+1: https://github.com/dimagi/commcare-hq/pull/24120/commits/41c19fd0c880e79742ab924911136e81cf463030

File: corehq/apps/app_manager/app_schemas/form_metadata.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: If it's not a big increase in scope, it'd be a good improvement. Checking whether `id` is index or unique id is something I do fairly often in app manager.

I wish we could do `id` vs `index`, but changing the model properties would be unreasonable, so using `id` and `unique_id` consistently everywhere would be the next best thing.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/app_manager/app_schemas/form_metadata.py (hunk 2 above)
Comment (proteusvacuum) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: :+1: I will do this in a separate PR

Comment (proteusvacuum) [reply corehq/apps/app_manager/app_schemas/form_metadata.py]: https://github.com/dimagi/commcare-hq/pull/24136

**Pull request merged**